### PR TITLE
Add fast-scroll indexer to MediaAdapter (Library View)

### PR DESCRIPTION
**Pull request by Adonai**
Add fast-scroller and implement on-cursor-change section indexer for MediaAdapter

Works flawlessly on Songs/Albums/Genres tabs

### CODE REVIEW

**Comment by adrian-bl** [comment]
Thanks for the PR.

Hmm.. we disabled fast-scroll back then via commit 8d9ff2823c480b7d05d026bedfcba5ba2e45e451 as it turned out to be VERY annoying on Android 5.0 - and it still seems to be on 5.1, at least on my Z3C :-/

See: https://webm.lol/v/1z5dc6d59c

It becomes very hard to hit the expand button as the expanded scroll bar seems to steal-away some touches, causing the navigation to become a nightmare.

I'd really love to merge this as having an index is a very nice feature - but we first need to find a way to make the fascroll-bar less 'greedy'. Enabling fast scroll currently makes Vanilla Music unusable on my Z3 :-/ 

@aminb You have a Nexus 5, correct? How bad was the FastScroll on 5.x on your device? (My Z3 is 4.6'')

**Comment by adrian-bl** [comment]
Looks like we hit the same issue as in http://stackoverflow.com/questions/33619453/scrollbar-touch-area-in-android-6

**Comment by Adonai** [comment]
> Hmm.. we disabled fast-scroll back then via commit 8d9ff28 as it turned out to be VERY annoying on Android 5.0 - and it still seems to be on 5.1, at least on my Z3C :-/

Why was fastscroll thumb always enabled back then? Autohiding option could reduce possible annoyance. 

From my point, I always find it annoying to scroll over my 17000-song playlist to find one that I would like to play now. Finding it in through search is not an option - too many clicks, takes too long.

Besides, former implementation didn't count non-latin letters and I always found myself skipping cyrillic half of my playlist.

**Comment by adrian-bl** [comment]
> Why was fastscroll thumb always enabled back then? Autohiding option could reduce possible annoyance. 

It wasn't an issue in Android 4.x: The HOLO theme doesn't have this annoying large touch area, so having the scrollbar always visible didn't cause any issues.

Auto-Hiding doesn't help much on Android 5.x (the linked video was recorded with your patch applied) as scrolling -> expanding is a common work flow - and expanding a row won't work well until the scroll bar hid itself :-/

> From my point, I always find it annoying to scroll over my 17000-song playlist to find one that I would 
> like to play now.

Yes, i'd also like to have it back - but as i said: it makes Vanilla Music very hard to use on my 4.7 inch phone.

> Besides, former implementation didn't count non-latin letters and I always found myself skipping cyrillic 
> half of my playlist.

I agree: The old implementation was overly-complicated and didn't work that well for non latin characters.

**Comment by Adonai** [comment]
Ok, can close this for now then

**Comment by aminb** [comment]
This pull request enabled fast-scroll only for my 'Songs' tab, not for the other ones where we have 'expand'.

However, I tried tapping in the right margin (where the expand buttons would be) and the fast-scroller captures the tap, unfortunately.

However, I'd really us to have fast-scroller enabled. So, maybe we could think about how we would change the 'expand' behaviour to have them both at the same time. Maybe we could switch to a "3-dot" menu and give it some padding from the right margin, or extend the scrollview and attempt to fix this behaviour of fast-scroll ?

**Comment by adrian-bl** [comment]
> This pull request enabled fast-scroll only for my 'Songs' tab, not for the other ones where we have 'expand' (like Albums or Genres).

That is a feature of Android: The fast-scroll will only get (automatically) enabled if there are more than 4  (i think) pages to scroll.

> Maybe we could switch to a "3-dot" menu and give it some padding from the right margin, or extend the scrollview and attempt to fix this behaviour of fast-scroll 

A 3-dot menu would look more material'ish, but i somehow really dislike it as i find long-press much more convinient.

But i noticed that CM-Music seems to suffer far less from the fast-scroll problem: i'll have a peek at the source to try to figure out what they are doing.

**Comment by Adonai** [comment]
Oh great, thanks for looking into it. Any help is needed from my side, just write here

File: src/ch/blinkenlights/android/vanilla/MediaAdapter.java [hunk 1]
```diff
@@ -635,4 +662,96 @@ public boolean hasStableIds()
...
+		// clip to start
+		if(sectionIndex < 0) {
+			return 0;
+		}
+		
+		// clip to end
+		if(sectionIndex >= mAlphabet.size()) {
+			return mCursor.getCount() - 1;
+		}
+
```
Comment (adrian-bl) [inline src/ch/blinkenlights/android/vanilla/MediaAdapter.java]: superfluent empty whitespace

File: src/ch/blinkenlights/android/vanilla/MediaAdapter.java [hunk 2]
```diff
@@ -635,4 +662,96 @@ public boolean hasStableIds()
...
+			}
+			if (!next.equals(lastKnown)) { // new char
+				mAlphabet.add(new SectionIndex(next, cursor.getPosition()));
+				lastKnown = next;
+			}
+		} while (cursor.moveToNext());
+	}
+	
+	@Override
+	public Object[] getSections()
```
Comment (adrian-bl) [inline src/ch/blinkenlights/android/vanilla/MediaAdapter.java]: whitespace at the end

**Comment by adrian-bl** [comment]
@Adonai Looks like a superfluent call to setScrollBarStyle (which always used outside_overlay) caused the issues.

I created a 'divider' branch (https://github.com/vanilla-music/vanilla/commits/divider) which fixes this and  also adds a left-padding to the divider on API >= 21. With this changes, fastScroll works ok on my Z3.

Could you please update the PR with a few changes before the merge?

* Please always add a boilerplate-comment to function definitions (purpose, arguments and return value)
* Please remove empty lines with whitespaces + whitespaces at the end of function definitions (see my comments in the PR, there are more of them: git diff on the console shows them in red)
* The 'Magic sort mode: sort by playcount' comment in 'MediaAdapter.java' is on the wrong place ;-)

**Comment by Adonai** [comment]
No problem, will update in a minute.
Thanks!

**Comment by adrian-bl** [comment]
Thanks for the updated pull request!

Shall i add you to the list of contributors (in Settings -> About) ?

**Comment by Adonai** [comment]
> Shall i add you to the list of contributors (in Settings -> About) ?

Too early, but if you're ok with this, why not :)

**Comment by adrian-bl** [comment]
Ok: What name should i add in the about.html file? (Adonai?)

Btw: I'm still not super happy with the fastScroll behaviour on small screens:
I dug trough the AOSP sources and figured out that FastScroller.java intercepts a hardcoded area of at least 48dip (defined in com.android.internal.R.dimen.fast_scroller_minimum_touch_target) and Android provides no mechanism to make this area smaller.

So i hacked up a small ListView class wich 'guards' this 48dip-wide area by rewriting the X-coordinates of touch events: https://github.com/vanilla-music/vanilla/commit/5b173748efb48325ba7792c0639cdd13c0330289

Its a rather hacky solution, but makes things much better on my Z3C - but not sure if we should really merge such a hack :-/

**Comment by Adonai** [comment]
>Ok: What name should i add in the about.html file? (Adonai?)

Oh sorry forgot to provide real name. Please put it as  `Oleg (Adonai) Chernovskiy <adonai@xaker.ru>`
> not sure if we should really merge such a hack :-/

I'm ok with this. Isolated in `layout-(l/m)dpi` folder and with appropriate comment about workaround it won't be a mess.

**Comment by adrian-bl** [comment]
I added you to the file via 2967bb38eb66a4fd0257c3824c59ec2ad4be106b

> I'm ok with this. Isolated in layout-(l/m)dpi folder and with appropriate comment about workaround it won't 
> be a mess.

It also adds a new class to rewrite touch events.
I pushed this into the master for now as it makes Vanilla much more useable on small screens - and, while it is not a great solution, it won't hurt us much.

**Pull request merged**